File: daily/cpp/2966_divide_array_into_arrays_w_max_diffs.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
    vector<vector<int>> divideArray(vector<int>& nums, int k) {
        int n = nums.size();
        if (n % 3 != 0) return {};
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        // for (int i = 0; i < n/3; ++i) {
        //     vector<int> tmp;
        //     for (int j = 0; j < 3; ++j) {
        //         if (nums[i*3+j] - nums[i*3] > k) return {};
        //         tmp.push_back(nums[i*3+j]);
        //     }
        //     res.push_back(tmp);
        // }
        for (int i = 0; i <= n - 3; i+=3) {
            if (nums[i+2] - nums[i] > k) return {};
            res.push_back({nums[i], nums[i+1], nums[i+2]});
        }
        return res;
    }
};

File: daily/rust/2966_divide_array_into_arrays_w_max_diffs.rs
impl Solution {
    pub fn divide_array(mut nums: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        /// This is fast, but not fast enough
        /// 
        // let mut nums = nums;
        // nums.sort();
        // let mut res = vec![];
        // for i in (0..=nums.len() - 3).step_by(3) {
        //     if nums[i + 2] - nums[i] > k {
        //         return vec![];
        //     } 
        //     res.push(vec![nums[i], nums[i + 1], nums[i + 2]]);
        // } 
        // res

        /// The following solution is way more faster
        /// 
        // nums.sort_unstable();
        // match nums.chunks(3).map(|c| c[2] - c[0]).any(|d| d > k) {
        //     true => vec![],
        //     false => nums.chunks(3).map(|c| c.to_vec()).collect(),
        // }

        nums.sort_unstable();
        if nums.chunks(3).map(|c| c[2] - c[0]).max().unwrap() > k {
            return vec![];
        }
        nums.chunks(3).map(|c| c.to_vec()).collect()
    }
}